### ISSUE TITLE
CacheSize - How to set right size?

### ISSUE THREAD

**Issue by taducquang** [quetzalcoatl/blagominer#17]
I know i should use power of two but how to determine the right size for its? CacheSize writes datas in RAM, let's say i have 32 GB RAM, which amount should i set for CacheSize?
Also if there's more details of how it works, i would love to learn what information do it need to save and why does it need to free spaces every times it begin read.
Thanks.

**Comment by quetzalcoatl** [comment]
It's a piece of code that I didn't write, and I only refactored it and moved around, so I can check/explain it only by reading and analyzing the code.

From what I could quickly gather:
- CacheSize is 16384 by default, is expressed in nonces (1n=64bytes)
- CacheSize2 is 262144 by default, is expressed in nonces
- ReadChunkSize, defaults to CacheSize if not specified, is expressed in nonces

"nonce" is the most basic piece of data read from plot files. It can't read less than that. I mean, it could, but no work, no deadline calculations can be done, unless it reads data for at least one full nonce. Hence no point in specifying anything that does not divide by noncesize (64 bytes), as it would only complicate the algirithms and make them need to rebuild nonces from partial reads.

So:
- 16384 means 1 megabyte of cache1 and readchunksize
- 262144 means 16 megs

Personally, I don't like this convention. I'd just say "give me about 16 megs" and let it align/round it down as needed. But so far, that's how it was since beginning.

ReadChunkSize specifies how large a block will be read from the plot file at any given time. It's directly passed to `ReadFile` and of course the OS may read less than asked for (i.e. reading close to end of file condition, etc).

CacheSize is used to build an in-memory cache of nonce data, read from plot files. Blagominer actually uses 2 identical blocks of cache memory of this size. That is, specifying CacheSize to default 16384nonces (1mb worth of data), makes Blagominer use **at least** 2 megabytes of cache (also see below re directories/threads). This is because Blagominer currently uses buffer-swapping technique: 1 buffer is being filled with data, other buffer is being processed by worker threads for calculating deadlines. Then, buffers are swapped: threads get new data, old data is overwritten with next chunk of data from plot files. It's a pretty simple concept, and implementation follows in simplicity: blago works in internal "rounds" (not related to coin mining rounds). One internal round starts two tasks: "read data to buffer A" and "calculate data from buffer B". Only once BOTH tasks are done, A-B buffers are swapped. This means that setting CacheSize to **very high** values may sound fun, but if you over do it, you may be losing time at the start of the race: when new coin round starts, blago starts reading data for new round, but has no data in buffer B to start calculating. It has to wait until buffer A fill. If that buffer is huge, deadline calculation will be **delayed** more. Note that it is **always** delayed until first filling first buffer finishes (for pretty obvious reasons). It's just a matter of how long the delay will be.

Anyways, always keep ReadChunkSize <= CacheSize, and since CacheSize block is filled with ReadChunkSize parts, it's nice to have it aligned, i.e. CacheSize=4xReadChunkSize, CacheSize=123xChunkSize seem fine, but not CacheSize=3.1415926xChunkSize. Even if you use this `π*ChunkSize`, it still should work. Nice multipliers feel like a good idea, but try, measure, see what works for you the best.

CacheSize2 seems related **only** to a theoretical PoC1<>PoC2 on the fly conversion and only for "PoC1 optimized plots". I'd say, this setting is ignorable. Noone uses PoC1 anymore. Maybe it would be relevant for some new coin whose creators forgot to kill PoC1 on the very start. Aside from that, all current coins use PoC2 mode and all reasonable new coins start off already in PoC2 mode. I was considering erasing that piece of code for quite a time. It's been obsolete even before I came to this project.

There's also a value called 'stagger'. CacheSize cannot be higher than stagger size, and it's automatically limited to that value. Stagger size is determined automatically from the filename, and IIRC, it is relevant for PoC1 only (see above) and PoC2 doesn't care about stagger anymore. Yet another deletium IMHO.

There's also BFS reading mode.. It's a cool thing, but I have no idea if anyone still uses it. So far I plan to keep it unless it makes something way too hard (and I see no chance for that happening). For no obvious reasons, even in PoC1 mode it uses CacheSize1 and ignores CacheSize2. Well, that just makes CacheSize2 even more ignorable.

Regarding requirement for "powers of two", I don't remember any of that. Right now it's already always a multiple of noncesize (64bytes), so adding more 2s to the multiplies will keep it a power of 2, yeah, true. But the SHABAL hash calculator uses its own chunk sizes, not configurable, platform-dependent, and reading files cares more about sector sizes than powers of two. Since CacheSize and ReadChunkSize are tightly related, it seems good to have them as multiplies of disk sector size. Traditional sector size is 512bytes, basic nonce is 64bytes, so it's 8 nonces per sector - so keep CacheSize and ReadChunkSize as multipliers of 8, otherwise you will be forcing your disk to read/seek somewhere in the middle of a sector and it's a bit wasteful. Also, disk cluster size may be involved in a similar way. 4kB? 8k? 32k? check your partitions. Maybe it will do some difference.

Regarding "what information do it need to save and why does it need to free spaces every times it begin read" - I don't fully understand, but at least the latter part "why does it need to free spaces" - no it doesn't, at least, not always. It allocates buffers only at start of the work (work.cpp, line 232, in work_i(), VirtualAlloc()) and frees them only on exit from work_i. This means that those buffers are then used in a buffer-swapping style through many rounds of reading files.

However, that's true that probably those buffers are now re-created a bit too often. The `work_i()` function represents a single worker that processes files within a single directory. This worker creates buffers only for its use. If you have 10 directories with plots, up to 10 workers may be created, all independent, all working in parallel, every of each processing its own files. (By the way, this means that default CacheSize=1mb + 100 plot directories may mean 200mb of cache memory allocated). Since workers are independent, there is not much point in mixing up their caches. That's why every worker gets its own cache. These workers work in parallel, so they all need their caches at the same time - so buffers are unlikely to sit unused and it's unlikely that those buffers could be shared or reused. Finally, in the event of having millions of directories and only few worker threads to chew through them all, yeah, cost of reallocating buffers could start to be noticeable.. but I think other more important problems would arise. So that's why it's not very tempting to aggressively reuse those buffers cross threads. As for historical reasons, these were simpler - PoC1 stagger. If you had PoC1 plots, then each directory could have its own different stagger value, which limits the size of CacheSize1, and theoretical possibility of having different cache sizes per directory makes the point. Of course, with PoC2 it doesn't matter anymore.

I have to admit, I have NOT ever measured how much time is wasted on re-creating those buffers. Since each pair of buffers is created once per network round, and once per-directory, and since it's usually giga- or terabytes per directory, I **think** the cost of buffer managing is neglible. With PoC2 everywhere, and no stagger thing anymore, I think it is possible to keep those buffers preallocated once and forever reused. But I don't know if it's worth spending time on that feature. 

Any measurements would be welcome :)

**Comment by taducquang** [comment]
Thanks quetzalcoatl, i read your reply and its really easy to understand.
VirtualAlloc did take paged size instead of RAM, when i set CacheSize to "1048576" and larger, the maximum RAM it takes is only around 7GB RAM (I have 20 hard drives) but the page size keep large in each settings until "failed to allocate memory".
I think i will use "lock pages on memory" feature in windows to measure how much RAM it actually takes and also to check if its faster than page size. Probably it can makes me to improve performance.

**Comment by quetzalcoatl** [comment]
If you'are trying to optimize that part, then yeah, RAM will be many times faster than any page file *), even if you have the page file on a SSD drive (though the best SSD drives are getting close in speed to RAM).

Interesting though about the virtual/physical difference. It's obvious, but it didn't occur to me earlier. Yes, in general, Blago uses VirtualAlloc and while the OS can assign physical, or pagefile memory, and while we often think it "gives physical until it runs out and then gives pagefile" it's not true, and the OS is free to choose whatever.

There's a [AllocateUserPhysicalPages](https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-allocateuserphysicalpages) funciton that I think goes in pair with [MEM_PHYSICAL flag](https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc#mem_physical) from VirtualAlloc and that should be able to guarantee allocation of physical memory. These caches need to be fast, so allocating them on the page file somewhat defeats the point, especially if the page file is on a non-SSD drive.

However, I'd need to experiment with this. You won't be able to configure that via config file right now. I'll try out and let you know if something works. In the meantime, do try out your ideas, maybe you'll find out somethign interesting.

> VirtualAlloc did take paged size instead of RAM, when i set CacheSize to "1048576" and larger, the maximum RAM it takes is only around 7GB RAM (I have 20 hard drives) but the page size keep large in each settings until "failed to allocate memory".

I didn't fully get what you meant here. Did you mean that you were testing it out and you were increasing the CacheSize until you got "failed to allocate" error? Or did you mean that the Blago has some leak and after you set CacheSize to a larger value, it kept old pages and kept growing on its own and eventually crashed with out of memory condition?

Also, do you have more than 7GB RAM available? If not, for example, if you have 8GB of RAM, then it sounds fine! 7GB for user process, 1GB for the OS to stay alive, and the rest of the allocation went from the pagefile. That's more or less how it was intended to should work.

Finally, 20 drives means at least 20 main worker processes. So 20 caches. So with current way Blago works, it's x20 multiplier for all memory requirements. 1M nonces = 64MB per cache = 128MB per worker, x20 = 2GB, I mean, at least. So with 8G of RAM probably it won't be of any sense to raise it much higher than 4M or 8M nonces.

And blago will try to run all 20 workers at the same time. Can your CPU support that many threads? I think it actually can go up to 2x or 3x that number (so 20-60 threads, depending on algorithm phase and luck). Unless your CPU is a multi-core-monster, most of the time most of those threads will sleep and most of their cache will be a dead weight. Switching contexts between 20-60 threads also may be a significant overhead. 

This threading model is ... bad. It's something I aim to rewrite almost from scratch in future versions, but that's not coming anytime very soon. Blago blindly starts a new main worker for every "directory" listed in the config file (blagominer.cpp, line ~1753, `size_t workersNeeded = ....  miningCoin->mining->dirs.size();`). Usually, that means for every drive. It's easy to loose performance just by adding too many sources to one Blago. Thread number rises, CPU starts wasting time cycling between threads. If that happens, it's better to **somehow** aggregate files into fewer directories (NTFS hardlinks need to be on the same drive, so wont do, but maybe NTFS softlinks? it should work IMHO).

EDIT:
*) uuh.. Okay, to be 100% correct, I think you **could** make a ramdisk, and then put the cachefile on that ramdisk, and that the pagefile would be practically as fast as RAM, but pagefile on a ramdisk sounds.. uh. I mean, why? what for? why not use the RAM directly?
